taf/auth_repo.py

```diff
@@ -266,15 +268,26 @@ def targets_at_revisions(self, *commits, target_repos=None, default_branch=None)
                 continue
             repositories_at_revision = repositories_at_revision["repositories"]
 
-            # get names of all targets roles defined in the current revision
-            with self.repository_at_revision(commit):
-                roles_at_revision = self.get_all_targets_roles()
+            current_metadata = self.list_files_at_revision(
+                commit, METADATA_DIRECTORY_NAME
+            )
+            new_files = [
+                metadata_path
+                for metadata_path in current_metadata
+                if metadata_path not in previous_metadata
+            ]
+            previous_metadata = current_metadata
+            if len(new_files):
+                with self.repository_at_revision(commit):
+                    roles_at_revision = self.get_all_targets_roles()
 
             for role_name in roles_at_revision:
                 # targets metadata files corresponding to the found roles must exist
-                targets_at_revision = self.get_json(
+                targets_at_revision = self.safely_get_json(
                     commit, get_role_metadata_path(role_name)
                 )
+                if targets_at_revision is None:
+                    continue
```

Comment:
Note: we don't support removing delegations yet, but leave this here in the future.